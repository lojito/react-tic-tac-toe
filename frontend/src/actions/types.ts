import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

export type CategoryAction = {
  type: "CHANGE_CATEGORY";
  category: Categories;
};

export type DisabledAction = {
  type: "CHANGE_DISABLED";
  disabled: boolean;
};

export type FirstAction = {
  type: "CHANGE_FIRST";
  first: FirstPlayer;
};

export type UserImageAction = {
  type: "CHANGE_USER_IMAGE";
  userImage: ImageName;
};

export type ComputerImageAction = {
  type: "CHANGE_COMPUTER_IMAGE";
  computerImage: ImageName;
};

export type PlayingLevelAction = {
  type: "CHANGE_PLAYING_LEVEL";
  level: PlayingLevel;
};

export type GameOverAction = {
  type: "CHANGE_GAME_OVER";
  over: boolean;
};

export type ResultAction = {
  type: "CHANGE_GAME_RESULT";
  result: GameResult;
};

export type WinnersSquaresAction = {
  type: "CHANGE_WINNERS_SQUARES";
  winners: SquareLocation[];
};

export type ActionUnknow = {
  type: "UNKNOWN";
};

export type Action =
  | CategoryAction
  | DisabledAction
  | FirstAction
  | UserImageAction
  | ComputerImageAction
  | PlayingLevelAction
  | GameOverAction
  | ResultAction
  | WinnersSquaresAction
  | ActionUnknow;
