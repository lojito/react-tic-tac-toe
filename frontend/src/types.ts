export type CategoryId = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

export type CategoryHavanaLandmarks = {
  id: 0;
  name: "CATEGORY_HAVANA_LANDMARKS";
  folder: "habana";
};

export type CategoryMontrealLandmarks = {
  id: 1;
  name: "CATEGORY_MONTREAL_LANDMARKS";
  folder: "montreal";
};

export type CategoryVancouverLandmarks = {
  id: 2;
  name: "CATEGORY_VANCOUVER_LANDMARKS";
  folder: "vancouver";
};

export type CategorySpainLandmarks = {
  id: 3;
  name: "CATEGORY_SPAIN_LANDMARKS";
  folder: "spain";
};

export type CategoryGermanyLandmarks = {
  id: 4;
  name: "CATEGORY_GERMANY_LANDMARKS";
  folder: "germany";
};

export type CategorySoccerPlayers = {
  id: 5;
  name: "CATEGORY_SOCCER_PLAYERS";
  folder: "soccer";
};

export type CategoryFruitsAndVegetables = {
  id: 6;
  name: "CATEGORY_FRUITS_AND_VEGETABLES";
  folder: "fruits";
};

export type CategoryAnimals = {
  id: 7;
  name: "CATEGORY_ANIMALS";
  folder: "animals";
};

export type CategoryPuppies = {
  id: 8;
  name: "CATEGORY_PUPPIES";
  folder: "puppies";
};

export type CategorySeinfeld = {
  id: 9;
  name: "CATEGORY_SEINFELD";
  folder: "seinfeld";
};

export type Categories =
  | CategoryAnimals
  | CategoryFruitsAndVegetables
  | CategoryGermanyLandmarks
  | CategoryHavanaLandmarks
  | CategoryMontrealLandmarks
  | CategoryPuppies
  | CategorySeinfeld
  | CategorySoccerPlayers
  | CategorySpainLandmarks
  | CategoryVancouverLandmarks;

export type ImageFolder =
  | "habana"
  | "montreal"
  | "vancouver"
  | "spain"
  | "germany"
  | "soccer"
  | "fruits"
  | "animals"
  | "puppies"
  | "seinfeld";

export type ImageName =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19;

export type FirstPlayer = Player.User | Player.Computer;

export enum PlayingLevel {
  Easy,
  Smart,
}

export enum Player {
  User,
  Computer,
  Nobody,
}

export type GameBoard = [
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player,
  Player
];

export type SquareBoard = {
  disabled: boolean;
  id: string;
  imagePath: string;
  className: "win" | "";
};

export enum SquareLocation {
  TopLeft,
  TopCenter,
  TopRight,
  MiddleLeft,
  MiddleCenter,
  MiddleRight,
  BottomLeft,
  BottomCenter,
  BottomRight,
}

export enum GameResult {
  Draw,
  Lost,
  Started,
  Won,
}

export type DBGame = {
  _id?: string;
  board: GameBoard;
  categoryId: CategoryId;
  first: FirstPlayer;
  userImage: ImageName;
  computerImage: ImageName;
  level: PlayingLevel;
  result: GameResult;
  winners: SquareLocation[];
  user?: string;
  createdAt?: string;
  updatedAt?: string;
  __v?: number;
};
