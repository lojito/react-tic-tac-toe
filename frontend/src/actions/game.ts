import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

import {
  CategoryAction,
  ComputerImageAction,
  DisabledAction,
  FirstAction,
  GameOverAction,
  PlayingLevelAction,
  ResultAction,
  UserImageAction,
  WinnersSquaresAction,
} from "./types";

export const CHANGE_CATEGORY = "CHANGE_CATEGORY";
export const CHANGE_DISABLED = "CHANGE_DISABLED";
export const CHANGE_FIRST = "CHANGE_FIRST";
export const CHANGE_IMAGES = "CHANGE_IMAGES";
export const CHANGE_USER_IMAGE = "CHANGE_USER_IMAGE";
export const CHANGE_COMPUTER_IMAGE = "CHANGE_COMPUTER_IMAGE";
export const CHANGE_PLAYING_LEVEL = "CHANGE_PLAYING_LEVEL";
export const CHANGE_GAME_OVER = "CHANGE_GAME_OVER";
export const CHANGE_GAME_RESULT = "CHANGE_GAME_RESULT";
export const CHANGE_WINNERS_SQUARES = "CHANGE_WINNERS_SQUARES";

export const changeCategory = (category: Categories): CategoryAction => {
  return {
    type: CHANGE_CATEGORY,
    category,
  };
};

export const changeDisabled = (disabled: boolean): DisabledAction => {
  return {
    type: CHANGE_DISABLED,
    disabled,
  };
};

export const changeFirst = (first: FirstPlayer): FirstAction => {
  return {
    type: CHANGE_FIRST,
    first,
  };
};

export const changeUserImage = (userImage: ImageName): UserImageAction => {
  return {
    type: CHANGE_USER_IMAGE,
    userImage,
  };
};

export const changeComputerImage = (
  computerImage: ImageName
): ComputerImageAction => {
  return {
    type: CHANGE_COMPUTER_IMAGE,
    computerImage,
  };
};

export const changeLevel = (level: PlayingLevel): PlayingLevelAction => {
  return {
    type: CHANGE_PLAYING_LEVEL,
    level,
  };
};

export const changeOver = (over: boolean): GameOverAction => {
  return {
    type: CHANGE_GAME_OVER,
    over,
  };
};

export const changeResult = (result: GameResult): ResultAction => {
  return {
    type: CHANGE_GAME_RESULT,
    result,
  };
};

export const changeWinners = (
  winners: SquareLocation[]
): WinnersSquaresAction => {
  return {
    type: CHANGE_WINNERS_SQUARES,
    winners,
  };
};
