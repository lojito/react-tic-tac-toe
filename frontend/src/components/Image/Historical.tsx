import React, { FC, useContext } from "react";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import { ImageFolder, ImageName } from "../../types";
import "./Image.scss";

interface Props {
  folder: ImageFolder;
  userImage: ImageName;
  computerImage: ImageName;
}

const Historical: FC<Props> = ({ folder, userImage, computerImage }) => {
  const {
    dictionary: { IMAGE_USER_LABEL, IMAGE_COMPUTER_LABEL },
  } = useContext(DictionaryContext);
  const url = `${process.env.PUBLIC_URL}/images`;

  return (
    <div className="images historical" data-testid="images-historical">
      <div className="user-image">
        <span>{`${IMAGE_USER_LABEL}: `}</span>
        <img
          className="image-token"
          src={`${url}/${folder}/${userImage}.jpg`}
          alt="user"
        />
      </div>

      <div className="computer-image">
        <span>{`${IMAGE_COMPUTER_LABEL}: `}</span>
        <img
          className="image-token"
          src={`${url}/${folder}/${computerImage}.jpg`}
          alt="computer"
        />
      </div>
    </div>
  );
};

export default Historical;
