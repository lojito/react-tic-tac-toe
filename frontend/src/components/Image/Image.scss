.images {
  margin: 10px 0;
  font-weight: bold;

  .user-image,
  .computer-image {
    display: inline-block;
    width: 50%;
  }

  .image-token {
    width: 40%;
    vertical-align: middle;
  }

  .refresh {
    width: 13%;
    margin: 0 1%;
    position: relative;
    vertical-align: middle;
    cursor: pointer;
  }

  .disabled {
    pointer-events: none;
  }
}
