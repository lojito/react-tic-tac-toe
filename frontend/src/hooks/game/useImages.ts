import { useCallback, useState } from "react";
import { ImageName } from "../../types";

const IMAGES_PER_CATEGORY = 20;

function useImages(user: ImageName = 0, computer: ImageName = 1) {
  const [images, setImages] = useState({ user, computer });

  const randomImage = useCallback(() => {
    let image = Math.floor(Math.random() * IMAGES_PER_CATEGORY);
    while (image === images.computer || image === images.user) {
      image = Math.floor(Math.random() * IMAGES_PER_CATEGORY);
    }
    return image as ImageName;
  }, [images.computer, images.user]);

  const handleRefreshImage = useCallback(
    (e) => {
      setImages({
        ...images,
        [e.target.id]: randomImage(),
      });
    },
    [images, randomImage]
  );

  if (images.user === images.computer) {
    setImages({
      ...images,
      user: randomImage(),
    });
  }

  return { images, handleRefreshImage };
}

export default useImages;
