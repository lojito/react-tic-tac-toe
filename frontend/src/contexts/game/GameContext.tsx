import React, { createContext } from "react";
import { Action } from "../../actions/types";
import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../../types";

export type State = {
  category: Categories;
  userImage: ImageName;
  computerImage: ImageName;
  first: FirstPlayer;
  level: PlayingLevel;
  disabled: boolean;
  over: boolean;
  result: GameResult;
  winners: SquareLocation[];
};

export const GameContext = createContext<{
  game: State;
  dispatch: React.Dispatch<Action>;
}>({ game: {} as State, dispatch: {} as React.Dispatch<Action> });
