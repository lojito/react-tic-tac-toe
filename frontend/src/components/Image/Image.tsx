import React, { FC, memo, useContext, useEffect } from "react";
import { changeComputerImage, changeUserImage } from "../../actions/game";
import { Action } from "../../actions/types";
import refreshImage from "../../assets/images/refresh.png";
import {
  DictionaryContext,
  DictionaryEntry,
} from "../../contexts/dictionary/DictionaryContext";
import { GameContext } from "../../contexts/game/GameContext";
import useImages from "../../hooks/game/useImages";
import { ImageFolder, ImageName } from "../../types";
import "./Image.scss";

interface Props {
  disabled: boolean;
  folder: ImageFolder;
  user: ImageName;
  computer: ImageName;
  dispatch: React.Dispatch<Action>;
  dictionary: DictionaryEntry;
}

const MemoizedImage: FC<Props> = memo(
  ({ disabled, folder, user, computer, dispatch, dictionary }) => {
    const url = `${process.env.PUBLIC_URL}/images`;
    const { images, handleRefreshImage } = useImages(user, computer);
    const { user: userImage, computer: computerImage } = images;
    const {
      IMAGE_USER_LABEL,
      IMAGE_USER_TOOLTIP,
      IMAGE_REFRESH_TOOLTIP,
      IMAGE_COMPUTER_LABEL,
      IMAGE_COMPUTER_TOOLTIP,
    } = dictionary;
    useEffect(() => {
      if (userImage !== user) {
        dispatch(changeUserImage(userImage));
      } else if (computerImage !== computer) {
        dispatch(changeComputerImage(computerImage));
      }
    }, [userImage, computerImage, user, computer, dispatch]);

    const imgClassName = `refresh ${disabled ? " disabled" : ""}`;

    return (
      <div className="images" data-testid="images">
        <div className="user-image">
          <span>{IMAGE_USER_LABEL}: </span>
          <img
            className="image-token"
            src={`${url}/${folder}/${userImage}.jpg`}
            alt="user"
            title={IMAGE_USER_TOOLTIP}
          />
          <img
            className={imgClassName}
            id="user"
            src={refreshImage}
            alt="refresh-user-img"
            onClick={(e) => handleRefreshImage(e)}
            title={IMAGE_REFRESH_TOOLTIP}
          />
        </div>

        <div className="computer-image">
          <span>{IMAGE_COMPUTER_LABEL}: </span>
          <img
            className="image-token"
            src={`${url}/${folder}/${computerImage}.jpg?`}
            alt="computer"
            title={IMAGE_COMPUTER_TOOLTIP}
          />
          <img
            className={imgClassName}
            id="computer"
            src={refreshImage}
            alt="refresh-computer-img"
            onClick={(e) => handleRefreshImage(e)}
            title={IMAGE_REFRESH_TOOLTIP}
          />
        </div>
      </div>
    );
  }
);

const Image: FC = () => {
  const { dictionary } = useContext(DictionaryContext);
  const { game, dispatch } = useContext(GameContext);
  const folder = game.category.folder;
  const user = game.userImage;
  const computer = game.computerImage;

  return (
    <MemoizedImage
      disabled={game.disabled}
      folder={folder}
      user={user}
      computer={computer}
      dispatch={dispatch}
      dictionary={dictionary}
    />
  );
};

export default Image;
