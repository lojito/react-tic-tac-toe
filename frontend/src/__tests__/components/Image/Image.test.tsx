import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Image from "../../../components/Image/Image";
import {
  DictionaryContext,
  DictionaryEntry,
} from "../../../contexts/dictionary/DictionaryContext";
import { GameContext, State } from "../../../contexts/game/GameContext";
import * as hooks from "../../../hooks/game/useImages";
import { ImageName } from "../../../types";

const mockHandleRefreshImage = jest.fn();

jest.mock("../../../hooks/game/useImages", () => ({
  __esModule: true,
  default: (user = 0, computer = 1) => ({
    images: { user, computer },
    handleRefreshImage: mockHandleRefreshImage,
  }),
}));

jest.mock("../../../actions/game", () => {
  return {
    changeUserImage: (image: ImageName) => {
      return {
        type: "CHANGE_USER_IMAGE",
        image,
      };
    },
    changeComputerImage: (image: ImageName) => {
      return {
        type: "CHANGE_COMPUTER_IMAGE",
        image,
      };
    },
  };
});

describe("Image", () => {
  const game = {
    category: {
      id: 0,
      folder: "habana",
      name: "CATEGORY_HAVANA_LANDMARKS",
    },
    userImage: 0,
    computerImage: 1,
    disabled: false,
  } as State;

  const dispatch = jest.fn();

  interface Props {
    disabled?: boolean;
  }

  const TestComponent = ({ disabled = false }: Props = {}) => {
    game.disabled = disabled;

    const mockDictionary = {
      IMAGE_USER_LABEL: "You",
      IMAGE_USER_TOOLTIP: "You will play with this image.",
      IMAGE_REFRESH_TOOLTIP: "Click to play with a different image.",
      IMAGE_COMPUTER_LABEL: "PC",
      IMAGE_COMPUTER_TOOLTIP: "The computer will play with this image.",
    } as DictionaryEntry;

    return (
      <DictionaryContext.Provider value={{ dictionary: mockDictionary }}>
        <GameContext.Provider value={{ game, dispatch }}>
          <Image />
        </GameContext.Provider>
      </DictionaryContext.Provider>
    );
  };

  const renderImages = ({ disabled = false }: Props = {}) => {
    const utils = render(<TestComponent disabled={disabled} />);

    const divImages = screen.getByTestId("images");
    const userImage = screen.getByRole("img", { name: "refresh-user-img" });
    const computerImage = screen.getByRole("img", {
      name: "refresh-computer-img",
    });

    return { ...utils, divImages, userImage, computerImage, game };
  };

  it("renders the Image component correctly", () => {
    const { asFragment } = renderImages();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the images when disabled is true", () => {
    const { divImages } = renderImages();

    expect(divImages).toBeInTheDocument();
  });

  it("should display the images when disabled is false", () => {
    const { divImages } = renderImages();

    expect(divImages).toBeInTheDocument();
  });

  it("should call the refresh image handler when clicking on the user image", async () => {
    const user = userEvent.setup();
    const { userImage } = renderImages();

    await user.click(userImage);

    expect(mockHandleRefreshImage).toHaveBeenCalled();
  });

  it("should call the refresh image handler when clicking on the computer image", async () => {
    const user = userEvent.setup();
    const { computerImage } = renderImages();

    await user.click(computerImage);

    expect(mockHandleRefreshImage).toHaveBeenCalled();
  });

  it("should call the dispatch function when clicking on the user image", async () => {
    const user = userEvent.setup();
    const { userImage, rerender } = renderImages();

    await user.click(userImage);

    jest.spyOn(hooks, "default").mockImplementationOnce(() => ({
      images: { user: 2, computer: 1 },
      handleRefreshImage: mockHandleRefreshImage,
    }));

    rerender(<TestComponent />);

    expect(dispatch).toHaveBeenCalled();
  });

  it("should call the dispatch function when clicking on the computer image", async () => {
    const user = userEvent.setup();
    const { computerImage, rerender } = renderImages();

    await user.click(computerImage);

    jest.spyOn(hooks, "default").mockImplementationOnce(() => ({
      images: { user: 0, computer: 2 },
      handleRefreshImage: mockHandleRefreshImage,
    }));

    rerender(<TestComponent />);

    expect(dispatch).toHaveBeenCalled();
  });

  it("should not call the refresh image handler when clicking on the user image and the game is disabled", async () => {
    const user = userEvent.setup();
    const disabled = true;
    const { userImage } = renderImages({ disabled });

    try {
      await user.click(userImage);
    } catch (err) {}

    expect(mockHandleRefreshImage).not.toHaveBeenCalled();
  });

  it("should not call the refresh image handler when clicking on the computer image and the game is disabled", async () => {
    const user = userEvent.setup();
    const disabled = true;
    const { computerImage } = renderImages({ disabled });

    try {
      await user.click(computerImage);
    } catch (err) {}

    expect(mockHandleRefreshImage).not.toHaveBeenCalled();
  });
});
