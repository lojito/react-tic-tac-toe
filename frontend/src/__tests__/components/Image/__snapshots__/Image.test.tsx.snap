// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders the Image component correctly 1`] = `
<DocumentFragment>
  <div
    class="images"
    data-testid="images"
  >
    <div
      class="user-image"
    >
      <span>
        You: 
      </span>
      <img
        alt="user"
        class="image-token"
        src="undefined/images/habana/0.jpg"
        title="You will play with this image."
      />
      <img
        alt="refresh-user-img"
        class="refresh "
        id="user"
        src="test-file-stub"
        title="Click to play with a different image."
      />
    </div>
    <div
      class="computer-image"
    >
      <span>
        PC: 
      </span>
      <img
        alt="computer"
        class="image-token"
        src="undefined/images/habana/1.jpg?"
        title="The computer will play with this image."
      />
      <img
        alt="refresh-computer-img"
        class="refresh "
        id="computer"
        src="test-file-stub"
        title="Click to play with a different image."
      />
    </div>
  </div>
</DocumentFragment>
`;
